Internal Ydoc error blocks further edits

Not really possible to reproduce consistently but the scenario was that I was removing a bunch of nodes and at one point got:

<img width="1036" height="706" alt="Image" src="https://github.com/user-attachments/assets/69f3fe79-204d-422e-a96f-419a02675745" />

From the on the project was unusable and always returning an error.
Noticed the corresponding stacktrace in the logs:
```
[WARN] [2025-11-19T00:01:36.253] [io.helidon.webserver.http1.Http1Connection] Internal server error
io.helidon.http.RequestException: Internal error
        at io.helidon.http@4.2.2/io.helidon.http.RequestException$Builder.build(RequestException.java:139)
        at io.helidon.webserver@4.2.2/io.helidon.webserver.http1.Http1Connection.handle(Http1Connection.java:249)
        at io.helidon.webserver@4.2.2/io.helidon.webserver.ConnectionHandler.run(ConnectionHandler.java:175)
        at io.helidon.common.task@4.2.2/io.helidon.common.task.InterruptableTask.call(InterruptableTask.java:47)
        at io.helidon.webserver@4.2.2/io.helidon.webserver.ThreadPerTaskExecutor$ThreadBoundFuture.run(ThreadPerTaskExecutor.java:239)
        at java.base/java.lang.VirtualThread.run(VirtualThread.java:456)
Caused by: java.lang.RuntimeException: java.lang.InterruptedException
        at org.enso.ydoc.polyfill/org.enso.ydoc.polyfill.web.WebSocket.lambda$execute$0(WebSocket.java:152)
        at io.helidon.webserver.websocket@4.2.2/io.helidon.webserver.websocket.WsRoute.listener(WsRoute.java:56)
        at io.helidon.webserver.websocket@4.2.2/io.helidon.webserver.websocket.WsUpgrader.upgrade(WsUpgrader.java:171)
        at io.helidon.webserver@4.2.2/io.helidon.webserver.http1.Http1Connection.handle(Http1Connection.java:193)
        ... 4 common frames omitted
Caused by: java.lang.InterruptedException: null
        at java.base/java.util.concurrent.FutureTask.awaitDone(FutureTask.java:482)
        at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:192)
        at org.enso.ydoc.polyfill/org.enso.ydoc.polyfill.web.WebSocket.lambda$execute$0(WebSocket.java:149)
        ... 7 common frames omitted
```

--- a/lib/java/ydoc-polyfill/src/main/java/org/enso/ydoc/polyfill/web/WebSocket.java
+++ b/lib/java/ydoc-polyfill/src/main/java/org/enso/ydoc/polyfill/web/WebSocket.java
@@ -142,17 +142,24 @@ final class WebSocket implements ProxyExecutable {
                     () -> {
                       var connectionFuture =
                           executor.submit(
-                              () -> handleConnect.execute().as(WebSocketConnection.class));
-
-                      WebSocketConnection connection;
-                      try {
-                        connection = connectionFuture.get();
-                      } catch (InterruptedException | ExecutionException e) {
-                        log.error("Connection error", e);
-                        throw new RuntimeException(e);
+                              () -> {
+                                var value = handleConnect.execute();
+                                return value.as(WebSocketConnection.class);
+                              });
+                      for (; ; ) {
+                        WebSocketConnection connection;
+                        try {
+                          log.trace("Waiting for connection");
+                          connection = connectionFuture.get();
+                        } catch (InterruptedException e) {
+                          log.debug("Interrupted", e);
+                          continue;
+                        } catch (ExecutionException e) {
+                          log.error("Connection error", e);
+                          throw new RuntimeException(e);
+                        }
+                        return connection;
                       }
-
-                      return connection;
                     });
 
         var httpRouting = HttpRouting.builder().route(Method.GET, "_health", () -> "OK");

--- a/lib/java/ydoc-polyfill/src/test/java/org/enso/ydoc/polyfill/web/WebSocketConnectionInterruptedTest.java
+++ b/lib/java/ydoc-polyfill/src/test/java/org/enso/ydoc/polyfill/web/WebSocketConnectionInterruptedTest.java
@@ -1,0 +1,207 @@
+package org.enso.ydoc.polyfill.web;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+import io.helidon.webclient.websocket.WsClient;
+import io.helidon.websocket.WsListener;
+import io.helidon.websocket.WsSession;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.BlockingDeque;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import org.graalvm.polyglot.HostAccess;
+import org.graalvm.polyglot.Value;
+import org.junit.Test;
+
+public class WebSocketConnectionInterruptedTest {
+
+  @Test
+  public void interruptThreadBeforeExecution() throws Exception {
+    var mockExecutor = new MockExecutor();
+    var ctx = WebEnvironment.createContext(HostAccess.ALL).build();
+    WebEnvironment.initialize(ctx, mockExecutor);
+    var server =
+        """
+        let ret = [ null ]
+        let wss = new WebSocketServer({host: 'localhost', port: 33455});
+        wss.onconnect = function() {
+          ret[0] = "Successfully connected";
+        };
+        wss.start();
+        ret
+        """;
+    var res = new Object[1];
+    mockExecutor.execute(
+        () -> {
+          res[0] = ctx.eval("js", server);
+        });
+
+    {
+      var cdl = new CountDownLatch(1);
+      WsClient.builder()
+          .build()
+          .connect(
+              "ws://localhost:33455",
+              new WsListener() {
+                @Override
+                public void onOpen(WsSession session) {
+                  cdl.countDown();
+                }
+              });
+      cdl.await();
+    }
+    {
+      var cdl = new CountDownLatch(1);
+      mockExecutor.execute(
+          () -> {
+            assertNotNull("Result set", res[0]);
+            assertTrue(res[0] instanceof Value);
+            var v = (Value) res[0];
+            assertTrue("It is an array", v.hasArrayElements());
+            assertEquals(1, v.getArraySize());
+            res[0] = v.getArrayElement(0).asString();
+            cdl.countDown();
+          });
+      cdl.await();
+      assertEquals("Successfully connected", res[0]);
+    }
+  }
+
+  private static final class MockExecutor implements ScheduledExecutorService, Runnable {
+    private final Thread thread;
+    private final BlockingDeque<Runnable> run = new LinkedBlockingDeque<>();
+
+    public MockExecutor() {
+      thread = Thread.ofPlatform().name("mock executor").start(this);
+    }
+
+    @Override
+    public void run() {
+      for (; ; ) {
+        Runnable r;
+        try {
+          r = run.takeFirst();
+        } catch (InterruptedException ex) {
+          ex.printStackTrace();
+          continue;
+        }
+        r.run();
+      }
+    }
+
+    @Override
+    public void execute(Runnable command) {
+      run.add(command);
+    }
+
+    @Override
+    public <T> Future<T> submit(Callable<T> clbl) {
+      var f = new CompletableFuture<T>();
+      execute(
+          () -> {
+            try {
+              var r = clbl.call();
+              f.complete(r);
+            } catch (Throwable t) {
+              f.completeExceptionally(t);
+            }
+          });
+      //
+      // this generates an InterruptedException as soon as
+      // someone calls `f.get()`
+      //
+      Thread.currentThread().interrupt();
+      return f;
+    }
+
+    @Override
+    public ScheduledFuture<?> schedule(Runnable r, long l, TimeUnit tu) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public <V> ScheduledFuture<V> schedule(Callable<V> clbl, long l, TimeUnit tu) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public ScheduledFuture<?> scheduleAtFixedRate(Runnable r, long l, long l1, TimeUnit tu) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public ScheduledFuture<?> scheduleWithFixedDelay(Runnable r, long l, long l1, TimeUnit tu) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void shutdown() {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public List<Runnable> shutdownNow() {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean isShutdown() {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean isTerminated() {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean awaitTermination(long l, TimeUnit tu) throws InterruptedException {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public <T> Future<T> submit(Runnable r, T t) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Future<?> submit(Runnable r) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public <T> List<Future<T>> invokeAll(Collection<? extends Callable<T>> tasks)
+        throws InterruptedException {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public <T> List<Future<T>> invokeAll(
+        Collection<? extends Callable<T>> clctn, long l, TimeUnit tu) throws InterruptedException {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public <T> T invokeAny(Collection<? extends Callable<T>> tasks)
+        throws InterruptedException, ExecutionException {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public <T> T invokeAny(Collection<? extends Callable<T>> clctn, long l, TimeUnit tu)
+        throws InterruptedException, ExecutionException, TimeoutException {
+      throw new UnsupportedOperationException();
+    }
+  }
+}

